### ISSUE TITLE
Getting a Selenium::WebDriver::Error::InvalidSessionIdError when upgrading rbenv

### ISSUE THREAD

**Issue by jmelki39** [appium/appium#16433]
## The problem

Briefly describe the issue you are experiencing (or the feature you want to see added to Appium). Tell us what you were trying to do and what happened instead. Remember, this is _not_ a place to ask questions. For that, go to http://discuss.appium.io!

## Environment

* Appium version (or git revision) that exhibits the issue: Desktop v1.21.0
* Last Appium version that did not exhibit the issue (if applicable): v.1.14.0 Desktop
* Desktop OS/version used to run Appium: mac OS Big Sur v11.6.1
* Node.js version (unless using Appium.app|exe): 12.8.1
* Npm or Yarn package manager: Homebrew
* Mobile platform/version under test: Android 11
* Real device or emulator/simulator: Real Device Samsung Galaxy S10+
* Appium CLI or Appium.app|exe: Appium.app

## Details

I have been running my Android mobile automation fine for a couple years using Appium Desktop v1.14.0, Ruby on Rails and Cucumber without updating any versions because I know its fragile and I don't want to spend time fixing everything once I upgrade to newer versions. I ended up running `brew upgrade rbenv ruby-build` and it upgraded my rbenv and ruby-build and now my automation has been failing. In my trouble shooting, I upgraded Appium to v1.21.0 and continue to get the same error I was seeing in v1.14.0. I am using rbenv 2.6.3 and `rbenv install' support: /usr/local/bin/rbenv-install (ruby-build 20220125)

## Link to Appium logs

(https://gist.github.com/jmelki39/77237d559d904a1300061f99e9dedb4d)

## Code To Reproduce Issue [ Good To Have ]

I'm not even sure what part of Code to put in here, but I think the issue is with my mobile driver? And this is how I have it setup in the code:

    def initialize_mobile_driver
      $driver = nil

      caps = get_capabilities

      $driver = Appium::Driver.new(caps, true)
      Appium.promote_appium_methods(Support::Screens)

      $driver.start_driver
    end

    def driver_quit
      $driver.driver_quit
    end

**Comment by mykola-mokhnach** [comment]
I don't observe any issues on the server side.
The session has been deleted first:

> [debug] [35m[W3C (0e9f2455)][39m Calling AppiumDriver.deleteSession() with args: ["0e9f2455-8aa3-453b-8063-32bb23e269fb"]

and then a comment to find a new element is sent to it:

> [info] [35m[HTTP][39m [90m[39m[info] [35m[HTTP][39m [37m-->[39m [37mPOST[39m [37m/wd/hub/session/0e9f2455-8aa3-453b-8063-32bb23e269fb/element[39m

So the exception in the log is expected. Make sure the session management on the client side is working properly

**Comment by jmelki39** [comment]
@mykola-mokhnach So I am able to run my first test case successfully, but then get the ```
A session is either terminated or not started (Selenium::WebDriver::Error::InvalidSessionIdError)
      NoSuchDriverError: A session is either terminated or not started
```
error once Appium goes to my next test. My 'Before' and 'After' functions worked fine in the past where I initialize the mobile driver in Before and then quit the driver in my After. Why now is this no longer working?

**Comment by KazuCocoa** [comment]
The warning is logged when you call `promote_appium_methods` multiple times to notify redundant `define_method` call since ruby lib 11.1.0.
https://github.com/appium/ruby_lib/blob/5683456f6c6f434d3fa1490a250d23acc3033729/lib/appium_lib/appium.rb#L205-L208

**Comment by jmelki39** [comment]
@KazuCocoa I was using Ruby rbenv 2.6.3 and appium desktop 1.14.0. Now Im on Ruby rbenv 3.1.0 and appium desktop 1.21.0.

Any ideas on my driver issue?

**Comment by jmelki39** [comment]
This is what my Before function looks like (it hasn't changed since upgrading):


    ```def initialize_mobile_driver
      $driver = nil

      caps = get_capabilities

      $driver = Appium::Driver.new(caps, true)
      Appium.promote_appium_methods(Support::Screens)

      $driver.start_driver
    end```

**Comment by KazuCocoa** [comment]
I mean the version of https://github.com/appium/ruby_lib, not Ruby version. Usually you manage Ruby's dependencies with Gemfile and Gemfile.lock, I believe.

**Comment by mykola-mokhnach** [comment]
> error once Appium goes to my next test. My 'Before' and 'After' functions worked fine in the past where I initialize the mobile driver in Before and then quit the driver in my After. Why now is this no longer working?

AFAIK a major security issue was fixed which basically was allowing to ignore the session identifier. Now it is being verified, so it is not possible anymore to send session commands if a wrong/expired/deleted session id is provided in the request path.

**Comment by jmelki39** [comment]
@mykola-mokhnach Im still confused about what I'm doing wrong. How am I sending a session command that is wrong/expired/deleted? Im stop and restart the drive each time a new test runs. What is recommended to do?

**Comment by jmelki39** [comment]
> I mean the version of https://github.com/appium/ruby_lib, not Ruby version. Usually you manage Ruby's dependencies with Gemfile and Gemfile.lock, I believe.

I don't see 'ruby_lib' at all listed in my Gemfile.lock

**Comment by KazuCocoa** [comment]
The package name is appium_lib as its readme. So, https://rubygems.org/gems/appium_lib is. I guess you've upgraded old version to latest? one when you upgraded your Ruby version. Then. I guess your global variable usage got break in a version from the old one to the new one. Then, it needs debug your code and ruby_lib.

In ruby_lib, https://github.com/appium/ruby_lib/blob/master/lib/appium_lib/driver.rb#L203 puts themselves as `$driver` in Appium::Driver's initializer. 

Your reported behavior is probably the previous driver's global variable instance was used after the new one. The new test session did not update the global variable's value from the previous one.

Personally, i'd recommend you to not use global variable as possible to restrict variable's scope.

**Comment by jmelki39** [comment]
@KazuCocoa Thank you! So in order to not use the global variable, do I just change       $driver = Appium::Driver.new(caps, true) to be 'false'?

`    def get_capabilities
      ENV['PLATFORM_NAME'] == 'ANDROID'
        file = "appium_android.txt"

      text_file = File.join(__dir__, "../support/#{file}")
      caps = Appium.load_appium_txt(file: text_file)
      caps[:caps][:deviceName] ||= ENV['DEVICE_NAME']
      caps[:caps][:platformName] ||= ENV['PLATFORM_NAME']
      caps[:caps][:platformVersion] ||= ENV['PLATFORM_VERSION']
      caps[:caps][:app] ||= ENV['APP']
      caps[:caps][:udid] ||= ENV['UDID']

      caps
    end

    def initialize_mobile_driver
      $driver = nil

      caps = get_capabilities

      $driver = Appium::Driver.new(caps, true)
      Appium.promote_appium_methods(Support::Screens)

      $driver.start_driver
    end`

**Comment by KazuCocoa** [comment]
> do I just change $driver = Appium::Driver.new(caps, true) to be 'false'?

Yes. Then, the appium_ruby lib no longer modifies `$driver` internally.

**Comment by KazuCocoa** [comment]
Then, `$driver` in your code might affect your test case.
Did you confirm what version of appium_lib did you use in your env was Ruby 2.6, and what version in Ruby 3.1env?

Probably you can compare appium_lib `10.6.0` with Ruby 2.6, and appium_lib `11.0` with Ruby 2.6 etc so that the issue was appium_lib or your other dependencies/Ruby versions.

**Comment by jmelki39** [comment]
> Then, `$driver` in your code might affect your test case. Did you confirm what version of appium_lib did you used in your env was Ruby 2.6, and what version in Ruby 3.1 env?
> 
> Probably you can compare appium_lib `10.6.0` with Ruby 2.6, and appium_lib `11.0` with Ruby 2.6 etc so that the issue was appium_lib or your other dependencies/Ruby versions.

I believe it was version 10.4.0 for both Ruby 2.6 and Ruby 3.1. Should I be updated to version 11 or 12 though? I try to run `gem update appium_lib` but it says is already up to date

**Comment by KazuCocoa** [comment]
> I believe it was version 10.4.0 for both Ruby 2.6 and Ruby 3.1

Then, perhaps it is something Ruby 3.1 env's issue to handle global variable.

**Comment by mykola-mokhnach** [comment]
@KazuCocoa Shall we move the issue elsewhere to ruby client tracker? It has nothing to do with the server functionality

**Comment by KazuCocoa** [comment]
yup, let me move
